internal/dot/graph.go

```diff
@@ -252,6 +274,59 @@ func (dg *Graph) addToGroup(r *Result, id CtorID) {
 	group.Results = append(group.Results, r)
 }
 
+// Prune removes elements from the graph that do not have failed results.
+// Removing elements that do not have failing results makes the graph easier to debug, since non-failing
+// nodes and edges can clutter the graph and don't help the user debug.
+func (dg *Graph) Prune() {
+	dg.pruneCtors()
+	dg.pruneGroups()
+}
+
+// Remove constructors from the graph that do not have failing Results.
+func (dg *Graph) pruneCtors() {
+	var prunedCtors []*Ctor
+	for _, c := range dg.Ctors {
+		if _, ok := dg.Failed.FailedCtors[c.ID]; ok {
+			prunedCtors = append(prunedCtors, c)
+		} else {
+			dg.pruneGroupResults(c)
+			delete(dg.ctorMap, c.ID)
+		}
+	}
+
+	dg.Ctors = prunedCtors
+}
+
+// Remove groups from the graph that do not have failing Results.
+func (dg *Graph) pruneGroups() {
+	var prunedGroups []*Group
+	for _, g := range dg.Groups {
+		if _, ok := dg.Failed.FailedGroups[groupKey{t: g.Type, group: g.Name}]; ok {
+			prunedGroups = append(prunedGroups, g)
+		} else {
+			delete(dg.groupMap, groupKey{t: g.Type, group: g.Name})
+		}
+	}
+
+	dg.Groups = prunedGroups
+}
+
+// PruneGroupResults removes constructor results that are still referenced in
+// their respective Group object.
+func (dg *Graph) pruneGroupResults(c *Ctor) {
+	for _, r := range c.Results {
+		var prunedResults []*Result
+		if resultGroup, ok := dg.groupMap[r.groupKey()]; ok {
```

Comment:
Result.groupKey will return `groupKey{Name: "", Type: ..}` when this isn't a group-result. You should probably check if `Group != ""` first.

Reply:
Also, code-style-wise, reduce nesting here. 

```
rg, ok := ...
if !ok {
    continue
}

for _, gr := rg.Results {
    ...
}

rg.Results = prunedResults
```

That can be simplified further by exploiting the fact that `foo[k]` returns
zero-value in the single-value form. So you can just do,

```
rg := dg.groupMap[..]
if rg == nil { continue }
```

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Reply:
I changed the code to integrate continue statements in. I don't believe there is much of a concision improvement of 

rg := dg.groupMap[...]
over
		rg, ok := dg.groupMap[k]
		if !ok {
			continue
		}